File: p014.py
def p014(limit):
    cache, maxcl, maxn = {}, 0, 0

    def collatz(i):
        cl = 0
        while i > 1:
            if i in cache:
                return cl + cache[i]
            else:
                if i % 2 == 0:
                    i /= 2
                else:
                    i = 3 * i + 1
                cl += 1
        return cl

    for i in xrange(limit):
        cache[i] = collatz(i)
        if cache[i] > maxcl:
            maxcl = cache[i]
            maxn = i

    return maxn

if __name__ == "__main__":
    from benchmark import benchmark
    benchmark(p014, 1000000)

File: p014.clj
(use '[euler :only [run]])

(def results (ref {}))

(defn add-to-cache [n c]
  (dosync
    (alter results assoc n c)))

(defn collatz
  ([n] (collatz n n 1))
  ([n n' c]
    (if (= 1 n')
      (do
        (add-to-cache n c)
        c)
      (if-let [r (@results n')]
        (let [c' (+ c r)]
          (add-to-cache n c')
          c')
        (recur
          n
          (if (odd? n')
            (inc (* 3 n'))
            (bit-shift-right n' 1))
          (inc c))))))

(defn p014 [n]
  (->>
    (map (fn [x] [x (collatz x)]) (range 1 n))
    (sort-by last)
    last
    first))

(run (p014 1e6))
